### DISCUSSION TITLE
Extended sql queries support.

### DISCUSSION THREAD

**Discussion by MATEUSZROGALSKI** [pocketbase/pocketbase#1732] [Q&A]
Currently I'm about to move out from firebase and decided to use pocketbase but with my limited (or nonexistent) knowledge of GO language and lack of proper documentation It's really difficult to use anything more than simple plain `SELECT` query. This limits usage of pocketbase greatly as we cannot rely on heavily "relationed" data for example in e-commerce, more complex management systems etc.

Imagine a scenario where you want to find warehouse where specified items are stored. Currently we have limited support for that as we have to list specified items via API call, accumulate all the necessary identifiers and relations into huge pile and from there sort out by the number of matching items in that warehouse.  For example:

```` JS
import PocketBase from 'pocketbase';
const pb = new PocketBase('http://127.0.0.1:8090');

const queryResult = await pb.collection('warehouse_item').getList(1, 1000, {
    filter: 'item_id = <first_item_id> || item_id = <second_item_id> || item_id = <third_item_id> ...'
});

// here we need to iterate through queryResult to group them by 'warehouse_id'
let map = {};
queryResult.items.map(item => {
    if (item.warehouse_id in map) {
        map[item.warehouse_id].count += 1;
        map[item.warehouse_id].item_ids.push(item.item_id);
    }
    else {
        map[item.warehouse_id] = {
            count: 1,
            item_ids: [item.item_id];
        }
    }
});
````

With the approach I've stated above we're able to find out what the warehouse ( or shop or whatever else we need ) that contains the most items we requested is. This, as we all can see is a bit problematic solution that can be solved on the database side with lot less time needed. ( not to mention we do not use client CPU time for that ).

An example of SQL that roughly does what was said above is this:

```` SQL
SELECT
    distinct(wi.warehouse_id) as WHID,
    COUNT(wi.*) as MATCHES
FROM WAREHOUSE_ITEM as wi
WHERE wi.item_id IN (1,3,4,5)
GROUP BY (WHID)
ORDER BY MATCHES DESC
````
So, to conclude this request: Will there be a possibility to invoke our own queries from the API side? Or could someone extend the documentation to help us figure out how the data is structured under the hood and how can we create our own routes that would help us achieve it?

Current state of the [documentation](https://pocketbase.io/docs/database/) ( for this topic ) is .. "informational" at best. It doesn't explain how records and relations are preserved and how to eventually begin writing custom queries. It mostly serves a purpose of redirect to libraries used and inform people that "something like this is kind of possible if you dig enough".

**Comment by ganigeorgiev** [comment]
**Aggregations, and custom `SELECT` queries in general, is in the [roadmap](https://github.com/orgs/pocketbase/projects/2) and is targeted for the next v0.13/v0.14 release.**

The documentation can definitely be improved and I'll consider updating the "Database" section in the near future.

> It doesn't explain how records and relations are preserved and how to eventually begin writing custom queries.

Each collection is represented as a plain SQL table with the schema fields as columns + some other system fields (eg. `id`, `created`, `updated`). Depending on the "Max select" option of your relation field, the relation value could be stored as a plain id string OR a serialized json array of ids (aka. the same as the JSON representation in the response).

Here is an example that register a new endpoint and return something similar to your query (_I haven't tested it_):

```go
package main

import (
	"log"
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/list"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/grouped-warehouse-items",
			Handler: func(c echo.Context) error {
				ids := c.QueryParams()["item_id"]

				result := []struct {
					Whid    string `db:"whid" json:"whid"`
					Matches int    `db:"matches" json:"matches"`
				}{}

				err := app.Dao().DB().
					Select(
						"distinct(wi.warehouse_id) as whid",
						"count(wi.*) as matches",
					).
					From("warehouse_item wi").
					Where(dbx.In(
						"wi.item_id",
						list.ToInterfaceSlice(ids)...,
					)).
					GroupBy("whid").
					OrderBy("matches desc").
					All(&result)

				if err != nil {
					apis.NewBadRequestError("Failed to fetch warehouse items", err)
				}

				return c.JSON(http.StatusOK, result)
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.ActivityLogger(app),
			},
		})

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by krystof018** [reply]
Is there any progress on custom queries via the API? I couldn't find anything in the documentation or elsewhere.

> **Aggregations, and custom `SELECT` queries in general, is in the [roadmap](https://github.com/orgs/pocketbase/projects/2) and is targeted for the next v0.13/v0.14 release.**

**Reply by ganigeorgiev** [reply]
@krystof018 PocketBase v0.13+ comes with new ["View collection"](https://pocketbase.io/docs/collections/#view-collection) type support.
You can create a view collection from the _"Admin UI > Collections > New collection > Type: View"_.

**Comment by MATEUSZROGALSKI** [comment]
Thank you for extremaly fast response. 

Given your example and explanation I would have only one question. Can I assume that every relation is based on `id` Field?

What I mean is, lets Imagine we have 3 tables: `warehouse`, `warehouse_item` and `item` where `warehouse_item` is just a joint table for one `warehouse` to mamy `item`. Everything created via web interface. Should my query be more like:

```` SQL
SELECT * from warehouse
JOIN warehouse_item ON warehouse.id = warehouse_item.id
JOIN item ON warehouse_item.id = item.id
...
````

**Reply by ganigeorgiev** [reply]
> Can I assume that every relation is based on id Field?

Yes, all relations are based on the `id` column of the related table.

I'm not sure that I understand the above query (_eg. why `warehouse_item.id` is both equal to `warehouse.id` and `item.id`; maybe it should be `warehouse_item.warehouse` and `warehouse_item.item`?_), but in all cases you can freely join the collections as regular SQL tables.

Additionally, you can also explore the db and its structure via the `sqlite3` cli tool (_it is most likely installed by default on most operating systems_), eg. `sqlite3 ./pb_data/data.db` or use a DB gui like DBeaver for example.

_As mentioned earlier, aggregations and custom queries are planned for the next release, so the above will be simplified in the near future._

**Reply by TomFreudenberg** [reply]
Hi @ganigeorgiev in case we have February 2024 I want to ask if there is any progress here or maybe I missed it.

I am looking heavily for a `group by` clause to enter in a custom SQL request.

E.g. we want to see the orders of the last days and months by categories. Doing this by a number of api requests ist boring.

Thanks for your feedback
Tom

**Reply by ganigeorgiev** [reply]
@TomFreudenberg If you are referring to the custom `SELECT` query web APIs support we already impmeneted this as ["View collection" type](https://pocketbase.io/docs/collections/#view-collection).

(_see also the above comment https://github.com/pocketbase/pocketbase/discussions/1732#discussioncomment-6290312_)

**Reply by TomFreudenberg** [reply]
@ganigeorgiev Thanks for your response, I missed that but it's exactly what I am looking for.

**Reply by TomFreudenberg** [reply]
@ganigeorgiev I run into (let's say) some limits using "View colections" when building a dashboard for our current application.

Therefor we need a number of SQL aggregations for about 50 states. In case that these are very similar views but sorted or divided a little different each, we have to create a view for any of it. This is pain in covering the code while we could do this very easy when using SQL.

The easiest and best option would be to use kind of dynamic view but created just by a web request.

So would it be (already) possible to use SQL SELECT statements via the web API?

That would be a great feature (for us currently).

---

2nd. We are able to access the sqlite db file by sqlite library in parallel. In case that this are only READS there is no issue with WAL. But we already had twice a situation when doing changes or selecting data by the sqlite3 CLI client in parallel to the running pocketbase instance, that our data file gets corrupted (indexes) and we had to rebuilt that data files from backup.

So we are not sure, if using the sqlite data file with another SQL library in parallel is a good idea.

Thanks for your feedback and help in advance
Tom

**Reply by ganigeorgiev** [reply]
@TomFreudenberg

1. It is difficult to recommend or suggest an alternative without understanding your use case. There is no support out of the box to execute arbitrary dynamic SQL statements using the web APIs as it would be unsafe. Even if it is added in the future it will be allowed only for admins. You can build it on your own if you need to by registering a custom route with Go or JS but again your description is too vague and lack details to help you with anything more specific.

2. Unless your use case fit in one of the described cases in [How to corrupt an SQLite database](https://www.sqlite.org/howtocorrupt.html), corrupting the db file is very unlikely to happen by a concurrent write operation. The most common error you should expect is SQLITE_BUSY or "database is locked" errors. If you are asking for my opinion whether operating on the database file from another process is a good idea - No, I wouldn't recommend it as it comes with too many edge cases that you'll have to handle yourself. If you really need to - you should weight the pros and cons of using PocketBase (or SQLite) in your project. SQLite is a inprocess database and concurrent/multiprocess access is not its best use case.

If the above doesn't help, please consider opening a new Q&A discussion with more details about your use case and the problematic query(ies).

**Reply by TomFreudenberg** [reply]
Thanks @ganigeorgiev 

The options for corruptions seems valid and maybe we hit one of those like file handle or lock.

So, yes - use different libraries and different threads on same data file seems really not a good idea.

I will think about custom routes but checking all the pros and cons in using pocketbase or sqlite or other db directly, I guess we are more supported in using something other.

Nevertheless, thanks for any feedback and keep on with the good job!

Tom

**Comment by hossxosc8107** [comment]
SQL language is great language, simple and efficient, very complex query logic can be done very easily by SQL language. Human keeps invent wheels over and over again, very complex wheels is still wheels.

**Comment by devjaw** [comment]
we would like to find a way to use plain SQL or ORM like drizzle, please support
